File: nodejs-app/delete.js
const client = require('./config');

const usersIndex = 'users';
const friendshipsIndex = 'friendships';

async function updateFriendship(userName) {
    try {
        const { hits } = await client.search({
            index: friendshipsIndex,
            body: {
                query: {
                    bool: {
                        should: [
                            { match: { user: userName } },
                            { match: { friends: userName } },
                        ],
                    },
                },
            },
        });

        const results = hits?.hits || [];
        if (results.length === 0) {
            console.log(`No friendships found for user "${userName}".`);
            return;
        }

        for (const result of results) {
            const docId = result._id;
            const friendship = result._source;

            if (friendship.user === userName) {
                await client.delete({
                    index: friendshipsIndex,
                    id: docId,
                    refresh: true,
                });
                console.log(`Friendship owned by "${userName}" deleted.`);
            } else {
                const updatedFriends = friendship.friends.filter((friend) => friend !== userName);
                await client.update({
                    index: friendshipsIndex,
                    id: docId,
                    doc: { friends: updatedFriends },
                    refresh: true,
                });
                console.log(`Friendship updated for document ID ${docId}, removed "${userName}" from friends.`);
            }
        }
    } catch (error) {
        console.error(`Error updating friendships for "${userName}":`, error.meta?.body || error);
    }
}

async function deleteUserByName(name) {
    try {
        const { hits } = await client.search({
            index: usersIndex,
            body: {
                query: {
                    match: { name },
                },
            },
        });

        const results = hits?.hits || [];
        if (results.length === 0) {
            console.log(`User with name "${name}" not found.`);
            return;
        }

        const docId = results[0]._id;

        const response = await client.delete({
            index: usersIndex,
            id: docId,
            refresh: true,
        });

        console.log(`User with name "${name}" deleted successfully:`, response);
    } catch (error) {
        console.error(`Error deleting user with Name ${name}:`, error.meta?.body || error);
    }
}

async function deleteUserAndFriendships(name) {
    await updateFriendship(name);
    await deleteUserByName(name);
}

(async () => {
    await deleteUserAndFriendships('Frank');
})();

File: go-app/example/delete.go
package example

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"github.com/elastic/go-elasticsearch/v8"
	"io"
	"log"
)

func DeleteUser(es *elasticsearch.Client, name string) {
	query := map[string]interface{}{
		"query": map[string]interface{}{
			"match": map[string]interface{}{
				"name": name,
			},
		},
	}
	jsonQuery, _ := json.Marshal(query)

	userResponse, err := es.Search(
		es.Search.WithIndex("users"),
		es.Search.WithBody(bytes.NewReader(jsonQuery)),
		es.Search.WithContext(context.Background()),
	)
	if err != nil {
		log.Fatalf("Error searching user: %s", err)
	}
	defer func(body io.ReadCloser) {
		_ = body.Close()
	}(userResponse.Body)

	var userResult map[string]interface{}
	if err := json.NewDecoder(userResponse.Body).Decode(&userResult); err != nil {
		log.Fatalf("Error decoding user response: %s", err)
	}

	userHits := userResult["hits"].(map[string]interface{})["hits"].([]interface{})
	if len(userHits) == 0 {
		fmt.Printf("User \"%s\" not found.\n", name)
		return
	}

	userDocID := userHits[0].(map[string]interface{})["_id"].(string)

	_, err = es.Delete(
		"users",
		userDocID,
		es.Delete.WithContext(context.Background()),
		es.Delete.WithRefresh("true"),
	)
	if err != nil {
		log.Fatalf("Error deleting user: %s", err)
	}

	friendshipResponse, err := es.Search(
		es.Search.WithIndex("friendships"),
		es.Search.WithBody(bytes.NewReader(jsonQuery)),
		es.Search.WithContext(context.Background()),
	)
	if err != nil {
		log.Fatalf("Error searching friendship: %s", err)
	}
	defer func(body io.ReadCloser) {
		_ = body.Close()
	}(friendshipResponse.Body)

	var friendshipResult map[string]interface{}
	if err := json.NewDecoder(friendshipResponse.Body).Decode(&friendshipResult); err != nil {
		log.Fatalf("Error decoding friendship response: %s", err)
	}

	friendshipHits := friendshipResult["hits"].(map[string]interface{})["hits"].([]interface{})
	if len(friendshipHits) > 0 {
		friendshipDocID := friendshipHits[0].(map[string]interface{})["_id"].(string)
		_, err = es.Delete(
			"friendships",
			friendshipDocID,
			es.Delete.WithContext(context.Background()),
			es.Delete.WithRefresh("true"),
		)
		if err != nil {
			log.Fatalf("Error deleting friendship: %s", err)
		}
	}

	fmt.Printf("User \"%s\" and their friendship removed successfully.\n", name)
}
